### PR TITLE
Switch to Zeus

### PR DESCRIPTION
**Pull request by htot**
This PR brings:
- switch to zeus
- kernel updates
- switch U-Boot to upstream (+ a view patches)
- build-in support for btrfs into U-Boot and kernel (in preparation for bringing in full btrfs support)
- drop booting from usb
- support complete image boot (kernel, rootfs) from sdhc

As always this builds 32b system. For 64b checkout zeus (this is what I am working on).

### CODE REVIEW

**Comment by andy-shev** [comment]
@htot Perhaps @xlla also would be interested in this PR and branch.

File: meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh [hunk 1]
```diff
@@ -6,14 +6,22 @@ set -euf -o pipefail
...
+readonly BOOT_PART="/dev/mmcblk0p7"
 
 # Check if already run before
 if [ -d "${GADGET_BASE_DIR}" ]; then
     echo "Already registered gadgets"
     exit 0
 fi
 
+# Check if there is an update partition
+if [ -n "`ls /dev/disk/by-partlabel | grep update`" ]; then
```
Comment (andy-shev) [inline meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh:19]: I guess it is easier to write like `if ... | grep -q ...; then` (i.o.w. no `test` is needed)

File: meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb [hunk 2]
```diff
@@ -89,3 +89,12 @@ IMAGE_INSTALL_append = " libgpiod"
...
 
 DEPENDS += "u-boot"
+
+ROOTFS_POSTPROCESS_COMMAND += "install_initrd; "
+
+install_initrd() {
+    bbnote "Adding initrd to image ${IMAGE_ROOTFS}"
+    install -d {IMAGE_ROOTFS}/boot
+    bbnote "from ${DEPLOY_DIR_IMAGE}/core-image-minimal-initramfs-edison.cpio.gz"
+    install -m 0755 ${DEPLOY_DIR_IMAGE}/core-image-minimal-initramfs-edison.cpio.gz ${IMAGE_ROOTFS}/boot/initrd
```
Comment (andy-shev) [inline meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb:99]: Why `0755`?

File: meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init [hunk 3]
```diff
@@ -121,6 +121,21 @@ boot_root() {
         return 1
     fi
 
+    # copy fstab
+    if [ -h /realroot/etc/fstab ] || [ -f /realroot/etc/fstab ]; then
```
Comment (andy-shev) [inline meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init:125]: Better w/o bashisms, i.e. `if [ -h ... -a -f ... ]; then`.

File: meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init [hunk 4]
```diff
@@ -103,14 +105,22 @@ boot_root() {
...
     for dir in `awk '/\/run\/media\/.* /{print $2}' /proc/mounts`; do
         umount ${dir}
     done
+
+    # Allow for identification of the real root even after boot
+    mkdir -p  /realroot
+    mount -n ${ROOTFLAGS} ${ROOT_DEVICE} /realroot
+
+    # check if init found
+    if [ -h /realroot/sbin/init ] || [ -f /realroot/sbin/init ]; then
```
Comment (andy-shev) [inline meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init:118]: Yeah, bashism :-)

File: meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh (hunk 1 above)
Comment (htot) [reply meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh:19]: I'll check this.

File: meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb (hunk 2 above)
Comment (htot) [reply meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb:99]: Really don't know. I copied from an example. But eventually it ends up on the image `edison-image-edison.hddimg` which installs into the boot partition. And that contains:
```
-rwxr-xr-x 1 root root 8536576 mei 30 18:41 bzImage
-rwxr-xr-x 1 root root 9553849 mei 30 18:41 initrd
...
```
So at least it's consistent with the kernel permissions :-) What would you prefer?

File: meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init (hunk 3 above)
Comment (htot) [reply meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init:125]: I'll consider.

File: meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init (hunk 4 above)
Comment (htot) [reply meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init:118]: I'm really no good with shell scripts. Just looked up examples from bash tutorials.

**Comment by htot** [comment]
We may find that @alext-mkrs has no time to review. But I have built and installed 64b on emmc and sdhc. And just did a fresh setup + build on 32b.

File: meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh (hunk 1 above)
Comment (htot) [reply meta-intel-edison-bsp/recipes-support/gadget/files/conf-gadget.sh:19]: I can change it to this one liner keeping the test but removing unnecessary processing:

`[ -h "/dev/disk/by-partlabel/update" ] && USBDISK=${UPDATE_PART} || USBDISK=${BOOT_PART}`

File: meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init (hunk 3 above)
Comment (htot) [reply meta-intel-edison-distro/recipes-core/initrdscripts/initramfs-framework/init:125]: I don't know what to do here. I read:

>  Binary -a and -o are inherently ambiguous. Use 'test EXPR1 && test EXPR2' or 'test EXPR1 || test EXPR2' instead.

and

> Note the shell logical primitives also benefit from short circuit
operation, which can be significant for file attribute tests.

How about changing to:
`if [[ -h /realroot/etc/fstab || -f /realroot/etc/fstab ]]; then`

I think this should work for bash, ash and ksh. Correct?

**Comment by htot** [comment]
Your comments addressed. Thanks for reviewing.

**Pull request merged**